### PR TITLE
POC: Time window instead of interval

### PR DESCRIPTION
**Pull request by ac-rappi**

### CODE REVIEW

**Review by enarbona — APPROVED** [review]

File: gobreaker.go [hunk 1]
```diff
@@ -52,28 +52,111 @@ type Counts struct {
...
+	mutex         *sync.Mutex
+	counts        map[int64]*Counts
 }
 
-func (c *Counts) onSuccess() {
-	c.TotalSuccesses++
-	c.ConsecutiveSuccesses++
-	c.ConsecutiveFailures = 0
+func (c *countsPerMoment) getOrCreateCount(now int64) *Counts {
+	counts, ok := c.counts[now]
```
Comment (fabioboterolyons) [inline gobreaker.go]: Podriamos cambiar
`counts, exists := c.counts[now]`

File: gobreaker.go [hunk 2]
```diff
@@ -52,28 +52,111 @@ type Counts struct {
...
+	if !ok {
+		counts = &Counts{}
+		c.counts[now] = counts
+	}
+	return counts
 }
+func (c *countsPerMoment) prune(now time.Time) {
+	oldKeys := []int64{}
```
Comment (fabioboterolyons) [inline gobreaker.go]: Yo cambiaria el nombre de oldKeys -> expiredKeys

File: gobreaker.go [hunk 3]
```diff
@@ -52,28 +52,111 @@ type Counts struct {
...
+	}
+}
+
+func (c *countsPerMoment) assembleCount(now time.Time) Counts {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	c.prune(now)
+
+	ret := Counts{}
```
Comment (fabioboterolyons) [inline gobreaker.go]: Yo renombraria ret -> timeWindowMetrics

Comment (fabioboterolyons) [inline gobreaker.go]: timeWindowMetrics := &Counts{}

File: gobreaker.go [hunk 4]
```diff
@@ -52,28 +52,111 @@ type Counts struct {
...
+
+	accumulateConsecutivesSuccesses := true
+	accumulateConsecutivesFailures := true
+
+	timeLowerBound := now.Unix() - c.secondsToKeep
+
+	for moment := now.Unix(); moment >= timeLowerBound; moment-- {
+		counts := c.counts[moment]
+		if counts != nil {
+			ret.Requests = ret.Requests + counts.Requests
```
Comment (fabioboterolyons) [inline gobreaker.go]: `timeWindowMetrics.Requests += counts.Requests`

**Review by fabioboterolyons — APPROVED** [review]

**Review by juanda645 — APPROVED** [review]

**Pull request merged**